services/pull/pull.go

```diff
@@ -838,3 +865,175 @@ func IsHeadEqualWithBranch(ctx context.Context, pr *issues_model.PullRequest, br
 	}
 	return baseCommit.HasPreviousCommit(headCommit.ID)
 }
+
+// GetCodeOwners returns the code owners configuration
+// Return empty slice if files missing
+// Return error on file system errors
+// We're trying to do the best we can when parsing a file.
+// Invalid lines are skipped. Non-existent users and groups too.
+func GetCodeOwners(ctx context.Context, repo *repo_model.Repository, branch string) ([]*CodeOwnerRule, []string, error) {
+	files := []string{"CODEOWNERS", "docs/CODEOWNERS", ".gitea/CODEOWNERS"}
+	gitRepo, closer, err := git.RepositoryFromContextOrOpen(ctx, repo.RepoPath())
+	if err != nil {
+		return nil, nil, err
+	}
+	defer closer.Close()
+
+	if !gitRepo.IsBranchExist(branch) {
+		return nil, nil, &git.ErrBranchNotExist{Name: branch}
+	}
+
+	commit, err := gitRepo.GetBranchCommit(branch)
+
+	var data string
+	for _, file := range files {
+		if blob, err := commit.GetBlobByPath(file); err == nil {
+			data, err = blob.GetBlobContent()
+			if err == nil {
+				break
+			}
+		}
+	}
+
+	rules, warnings := GetCodeOwnersFromContent(ctx, data)
+	return rules, warnings, nil
+}
+
+func GetCodeOwnersFromContent(ctx context.Context, data string) ([]*CodeOwnerRule, []string) {
+	if len(data) == 0 {
+		return nil, nil
+	}
+
+	rules := make([]*CodeOwnerRule, 0)
+	lines := strings.Split(data, "\n")
+	warnings := make([]string, 0)
+
+	for i, line := range lines {
+		tokens := tokenizeCodeOwnersLine(line)
+		if len(tokens) == 0 {
+			continue
+		} else if len(tokens) < 2 {
+			warnings = append(warnings, fmt.Sprintf("Line: %d: incorrect format", i+1))
+			continue
+		}
+		rule, wr := parseCodeOwnersLine(ctx, tokens)
+		for _, w := range wr {
+			warnings = append(warnings, fmt.Sprintf("Line: %d: %s", i+1, w))
+		}
+		if rule == nil {
+			continue
+		}
+
+		rules = append(rules, rule)
+	}
+
+	return rules, warnings
+}
+
+type CodeOwnerRule struct {
+	Rule     *regexp.Regexp
+	Negative bool
+	Users    []*user_model.User
+}
+
+func parseCodeOwnersLine(ctx context.Context, tokens []string) (*CodeOwnerRule, []string) {
```

Comment:
IMO it needs some tests to cover the code, to make sure the logic won't be broken by future changes.